pkg/api/util/names.go

```diff
@@ -44,15 +44,65 @@ func ComputeName(prefix string, obj interface{}) (string, error) {
 	// and pods down the road for ACME resources.
 	prefix = DNSSafeShortenTo52Characters(prefix)
 
+	// the prefix is <= 52 characters, the decimal representation of
+	// the hash is <= 10 characters, and the hyphen is 1 character.
+	// 52 + 10 + 1 = 63, so we're good.
 	return fmt.Sprintf("%s-%d", prefix, hashF.Sum32()), nil
 }
 
-// DNSSafeShortenTo52Characters shortens the input string to 52 chars and ensures the last char is an alpha-numeric character.
-func DNSSafeShortenTo52Characters(in string) string {
-	if len(in) >= 52 {
-		validCharIndexes := regexp.MustCompile(`[a-zA-Z\d]`).FindAllStringIndex(fmt.Sprintf("%.52s", in), -1)
-		in = in[:validCharIndexes[len(validCharIndexes)-1][1]]
+// ComputeSecureUniqueDeterministicNameFromData computes a deterministic name from the given data.
+// The algorithm in use is SHA256 and is cryptographically secure.
+// The output is a string that is safe to use as a DNS label.
+// The output is guaranteed to be unique for the given input.
+// The output will be at least 64 characters long.
+func ComputeSecureUniqueDeterministicNameFromData(fullName string, maxNameLength int) (string, error) {
+	const hashLength = 64
+	if maxNameLength < hashLength {
+		return "", fmt.Errorf("maxNameLength must be at least %d", hashLength)
+	}
+
+	if len(fullName) <= maxNameLength {
+		return fullName, nil
+	}
+
+	hash := sha256.New()
+
+	_, err := hash.Write([]byte(fullName))
+	if err != nil {
+		return "", err
+	}
+
+	prefix := DNSSafeShortenToNCharacters(fullName, maxNameLength-hashLength-1)
+	hashResult := hash.Sum(nil)
+
+	if len(prefix) == 0 {
+		return fmt.Sprintf("%08x", hashResult), nil
 	}
 
-	return in
+	return fmt.Sprintf("%s-%08x", prefix, hashResult), nil
+}
+
+// DNSSafeShortenToNCharacters shortens the input string to N chars and ensures the last char is an alpha-numeric character.
+func DNSSafeShortenToNCharacters(in string, maxLength int) string {
```

Comment:
It seems the description doesn't match the behavior. It not only makes sure that the last char is an alphanumeric char, but it also removes some characters like `-` and `.`. Can the comment reflect that?

Also, why are `.` and `-` being removed?

Reply:
The function description is correct, we only remove all trailing non-alphanumerical characters.

Reply:
eg.:
```
DNSSafeShortenToNCharacters("a*aa.....", 3) == "a*a"
DNSSafeShortenToNCharacters("a**aa.....", 3) == "a" // "a**" without trailing non-alphanumerics
```

Reply:
My bad, I thought by reading the test cases that the `.` and `-` were removed anywhere in the string. I see that only the ending `.` and `-` are removed.

Reply:
These test cases are really looking for edge cases, making them a bit confusing sometimes.

Reply:
Why do we need to make sure that the last char is a non-alphanumerical character? Is that a restriction on `generateName`?

**Update:** yes, `generateName` requires a DNS label:

```console
$ k create -f- <<EOF                                                    
apiVersion: v1
kind: Secret
metadata:
  generateName: super-mega-long-certificate-name-that-uses-sixty-four-cha.
EOF
```

results in:

> The Secret `super-mega-long-certificate-name-that-uses-sixty-four-cha.bzjls` is invalid: metadata.generateName: Invalid value: `super-mega-long-certificate-name-that-uses-sixty-four-cha.`: a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')

No idea why the Kubernetes team chose to restrict `generateName` to a DNS label...

I'd suggest explaining why this function exists in the first place.